src/clients/http.rs

```diff
@@ -271,23 +291,191 @@ impl HttpClient {
+                response_status = response.status().as_u16(),
+                request_duration = latency.as_millis()
+            );
+        } else if response.status().is_client_error() {
+            // On every client error (HTTP 4xx) response
+            info!(
+                monotonic_counter.client_4xx_response_count = 1,
+                response_status = response.status().as_u16(),
+                request_duration = latency.as_millis()
+            );
+        } else if response.status().is_success() {
+            // On every successful (HTTP 2xx) response
+            info!(
+                monotonic_counter.client_success_response_count = 1,
+                response_status = response.status().as_u16(),
+                request_duration = latency.as_millis()
+            );
+        } else {
+            error!(
+                "unexpected HTTP client response status code: {}",
+                response.status().as_u16()
+            );
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ClientOnFailure;
+
+impl OnFailure<ServerErrorsFailureClass> for ClientOnFailure {
+    fn on_failure(
+        &mut self,
+        failure_classification: ServerErrorsFailureClass,
+        latency: Duration,
+        span: &Span,
+    ) {
+        let _guard = span.enter();
+        let trace_id = trace::current_trace_id();
+        let latency_ms = latency.as_millis().to_string();
+
+        let (status_code, error) = match failure_classification {
+            ServerErrorsFailureClass::StatusCode(status_code) => {
+                error!(
+                    ?trace_id,
+                    ?status_code,
+                    latency_ms,
+                    "failure handling request",
+                );
+                (Some(status_code), None)
+            }
+            ServerErrorsFailureClass::Error(error) => {
+                error!(?trace_id, latency_ms, "failure handling request: {}", error,);
+                (None, Some(error))
+            }
+        };
+
+        info!(
+            monotonic_counter.client_request_failure_count = 1,
+            latency_ms,
+            ?status_code,
+            ?error
+        );
+        info!(
+            monotonic_counter.client_5xx_response_count = 1,
+            latency_ms,
+            ?status_code,
+            ?error
+        );
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ClientOnEos;
+
+impl OnEos for ClientOnEos {
+    fn on_eos(self, _trailers: Option<&HeaderMap>, stream_duration: Duration, span: &Span) {
+        let _guard = span.enter();
+        info!(
+            trace_id = %trace::current_trace_id(),
+            duration_ms = stream_duration.as_millis(),
+            "end of stream",
+        );
+        info!(
+            monotonic_counter.client_stream_response_count = 1,
+            stream_duration = stream_duration.as_millis()
+        );
+        info!(histogram.client_stream_response_duration = stream_duration.as_millis() as u64);
```

Comment:
Duplicated fields, similar to comments above. I believe the counters can be moved up to the initial trace.